### ISSUE TITLE
Can't hide window decorations through qt5

### ISSUE THREAD

**Issue by daniwenne** [labwc/labwc#594]
Here is a test program I made, using Qt window flags to hide the title bar:
```
import sys
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtCore import Qt

class Window(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowFlag(Qt.FramelessWindowHint)
        self.setGeometry(100, 100, 400, 300)
        self.show()

App = QApplication(sys.argv)
window = Window()
sys.exit(App.exec())
```
If I run this on plasma or weston I get the expected behavior:
![plasma](https://user-images.githubusercontent.com/75865749/197311444-078d9678-eaba-4972-b52d-19e223baa084.png)

But on labwc I get this:
![image](https://user-images.githubusercontent.com/75865749/197311491-151564d9-d132-4914-8395-a0f3f78132e6.png)

I can still manually hide the decoration by using the title bar menu.

**Comment by johanmalm** [comment]
I think it relates back to https://github.com/labwc/labwc/issues/297
We need to get this sorted really :smile:

**Comment by heroin-moose** [comment]
Yeah, my issue exactly!

**Comment by johanmalm** [comment]
Okay. Here comes some analysis on this topic and a proposal.

This relates to issue #297 and #594

# References

- https://wayland.app/protocols/xdg-decoration-unstable-v1
- https://wayland.app/protocols/kde-server-decoration

# Current implementation

## xwayland

- SSD can be changed by the following:

  - XWayland surface signal `set_decorations` (Used by steam) and then
    determined by xwayland_surface->decorations which can be
    WLR_XWAYLAND_SURFACE_DECORATIONS_{ALL,NO_BORDER,NO_TITLE}
    We currently respect `ALL`. This seems to work and I cannot see any
    use-case for a user-override.

  - `ToggleDecorations` action which toggles SSD on/off.

## xdg-shell

- Decoration can be changed by the following:

  - rc.xml <core><decoration> values `server|client`

  - Protocol `xdg-toplevel-decoration` signal `request_mode`.
    Requesting values `SERVER|CLIENT|NONE`.
    Note: `NONE` means that the client does not care rather than no deco.
    As @dnkl said, we currently obey client `request_mode` wishes for
    `CLIENT` and `SERVER`.
    We respond to `NONE` (as a result of `unset_mode()`) by setting whatever
    `<core><decoration>` is set to (which is the case with `alacritty`).

  - KDE protocol `server-decoration`. We support this primarily (I think)
    because GTK does not (yet) support the xdg-toplevel-decoration protocol.
    https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/2191
    Interestingly it also supports `SERVER|CLIENT|NONE`, but `NONE` means
    no-decoration rather than no-opinion. 

  - `ToggleDecorations` action which toggles SSD on/off.

# The bug relating to foot

We check the outcome of KDE `server-decoration` protocol CSD/SSD negotiation
using a dubious check (https://github.com/labwc/labwc/blob/deb658b672fbf4810de33fb66c973aec20a3ebc6/src/xdg.c#L25-L33)
Without this foot works fine with `rc.xml` set to `client`. Whilst the check
has worked well in almost all circumstances, it is a stupid work-around
and we should implement this sort of thing: https://github.com/swaywm/sway/blob/master/sway/decoration.c
in order to determine the result of the SSD/CSD negotiation.

# The bug relating to the qt5 script in issue #594

Removing the `has_ssd()` check described above and setting `rc.xml` to
client makes the qt script work
(i.e. `self.setWindowFlag(Qt.FramelessWindowHint)` hides the CSD).
However, there is no way to specify no-decoration through the
`xdg-toplevel-decoration` protocol, other than for the client to request
CSD and then not show it (but the script does not seem to do that so
that is all a bit hypothetical). Implementing the KDE protocol fix described
above might also help us determine if a qt application wants 'no-decoration'
whatsoever. Not sure. I suppose that would give us another dependency on
that protocol.

# Proposal

1. Ignore client wishes when dealing with `xdg-toplevel-decoration`
   mode-requests. Easy to implement.
   https://github.com/labwc/labwc/blob/deb658b672fbf4810de33fb66c973aec20a3ebc6/src/xdg-deco.c#L24
   If we really want to honour set_mode() requests we could easily add a
   config option for that later if there is a use-case.

2. Implement https://github.com/swaywm/sway/blob/master/sway/decoration.c to
   correctly check the outcome of `server-decoration` negotiations.

# Side note

[`wlr_server_decorations.h`](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/include/wlr/types/wlr_server_decoration.h#L2)
warns that the `server-decoration` protocol is obsolete and will be removed
in future versions and that the recommended replacement is `xdg-decoration`.
I guess that timing depends on GTK and Qt implementations.

**Comment by Consolatis** [comment]
There were reports via IRC about some Gtk4 applications having the same issue, so I guess Gtk4 should be added to the test matrix as well.

**Comment by Consolatis** [comment]
Should be fixed by #838

Feel free to comment if that is not the case and we'll reopen the issue.
